next/src/visibility.ts

```diff
@@ -0,0 +1,163 @@
+        fieldSchema as JsfObjectSchema,
+        options,
+      )
+    }
+  }
+}
+
+/**
+ * Applies JSON Schema conditional rules to determine field visibility
+ * @param fields - The fields to apply rules to
+ * @param values - The current form values
+ * @param schema - The JSON schema containing the rules
+ * @param options - Validation options
+ *
+ * Fields start with visibility based on their required status.
+ * Conditional rules in the schema's allOf property can then:
+ * - Make fields visible by including them in a required array
+ * - Make fields hidden by setting them to false in properties
+ */
+function applySchemaRules(
+  fields: Field[],
+  values: SchemaValue,
+  schema: JsfObjectSchema,
+  options: ValidationOptions = {},
+) {
+  if (!schema.allOf || !Array.isArray(schema.allOf) || !isObjectValue(values)) {
+    return
+  }
+
+  const conditionalRules = schema.allOf
+    .filter(rule => typeof rule === 'object' && rule !== null && 'if' in rule)
+    .map((rule) => {
+      const ruleObj = rule as NonBooleanJsfSchema
+
+      const ifErrors = validateSchema(values, ruleObj.if!, options)
+      const matches = ifErrors.length === 0
+
+      // Prevent fields from being shown when required fields have type errors
+      let hasTypeErrors = false
+      if (matches
+        && typeof ruleObj.if === 'object'
+        && ruleObj.if !== null
+        && Array.isArray(ruleObj.if.required)) {
+        const requiredFields = ruleObj.if.required
+        hasTypeErrors = requiredFields.some((fieldName) => {
+          if (!schema.properties || !schema.properties[fieldName]) {
+            return false
+          }
+          const fieldSchema = schema.properties[fieldName]
+          const fieldValue = values[fieldName]
+          const fieldErrors = validateSchema(fieldValue, fieldSchema, options)
+          return fieldErrors.some(error => error.validation === 'type')
+        })
+      }
+
+      return { rule: ruleObj, matches: matches && !hasTypeErrors }
+    })
+
+  for (const field of fields) {
+    // Default visibility is based on required status
+    let isVisible = field.required
+
+    for (const { rule, matches } of conditionalRules) {
+      if (matches && rule.then) {
+        isVisible = isFieldVisible(field.name, rule.then, isVisible)
+      }
+      else if (!matches && rule.else) {
+        isVisible = isFieldVisible(field.name, rule.else, isVisible)
+      }
+    }
+
+    field.isVisible = isVisible
+  }
+}
+
+/**
+ * Determines whether a field should be visible based on a schema branch
+ * @param fieldName - The name of the field
+ * @param branch - The schema clause (either 'then' or 'else')
+ * @param currentVisibility - The current visibility state
+ * @returns The updated visibility state
+ *
+ * Visibility logic:
+ * - If the field is in the required array → make visible
+ * - If the field is explicitly false in properties → make hidden
+ * - Otherwise, preserve current visibility
+ */
+function isFieldVisible(
```

Comment:
The way it should work:
* a field is always visible unless the schema is `false`, or in this case the taken if branch.
* when both `const` and `default` are set on a schema it sets forcedValue and hides it but does not set isVisible to false